### PR TITLE
TPM1.2: Add NVWriteValue

### PR DESCRIPTION
**Pull request by ChriMarMe**
Add function to write NVRAM

### CODE REVIEW

**Comment by ChriMarMe** [comment]
@tmroeder and here

**Comment by ChriMarMe** [comment]
Really need some review here please

File: tpm/tpm.go [hunk 1]
```diff
@@ -929,6 +929,30 @@ func NVReadValue(rw io.ReadWriter, index, offset, len uint32, ownAuth digest) ([
 	return data, nil
 }
 
+func NVWriteValue(rw io.ReadWriter, index, offset uint32, data []byte, ownAuth digest) ([]byte, error) {
```
Comment (tmroeder) [inline tpm/tpm.go]: It looks like this function always returns nil in its first return value. So, please return only `error`

File: tpm/constants.go [hunk 2]
```diff
@@ -53,7 +53,10 @@ const (
 	ordOwnerReadInternalPub uint32 = 0x00000081
 	ordFlushSpecific        uint32 = 0x000000BA
 	ordPcrReset             uint32 = 0x000000C8
+	ordNVWriteValue         uint32 = 0x000000CD
+	ordNVWriteValueAuth     uint32 = 0x000000CE
```
Comment (tmroeder) [inline tpm/constants.go]: Please remove ordNV{Write,Read}ValueAuth, since neither is used in this commit.

File: tpm/tpm.go (hunk 1 above)
Comment (ChriMarMe) [reply tpm/tpm.go]: Whoopsi

**Comment by ChriMarMe** [comment]
Can't write a test which test this function unless GetNVList of #128 is merged. 
I need to read some proper addresses first where i can try to write to.

Otherwise I need to implement DefineSpace at first, so I can test all the together.
-define space at NVRAM 
-write a defined value
-read the value and compare to defined value
-release space at NVRAM for clean up

File: tpm/commands.go [hunk 3]
```diff
@@ -177,6 +177,24 @@ func nvReadValue(rw io.ReadWriter, index, offset, len uint32, ca *commandAuth) (
 	return b, &ra, ret, nil
 }
 
+// nvWriteValue writes to the NVRAM
+// If TPM isn't locked, no authentification is needed.
```
Comment (twitchy-jsonp) [inline tpm/commands.go]: Typo: `authentification` --> `authentication`

File: tpm/tpm.go [hunk 4]
```diff
@@ -940,6 +940,39 @@ func NVReadValue(rw io.ReadWriter, index, offset, len uint32, ownAuth []byte) ([
...
+	data, ra, ret, err := nvWriteValue(rw, index, offset, uint32(len(data)), data, ca)
+	if err != nil {
+		return fmt.Errorf("failed to read from NVRAM: %v", err)
+	}
+	raIn := []interface{}{ret, ordNVReadValue, tpmutil.U32Bytes(data)}
+	if err := ra.verify(ca.NonceOdd, sharedSecretOwn[:], raIn); err != nil {
+		return fmt.Errorf("failed to verify authenticity of response: %v", err)
+	}
+
+	return err
```
Comment (twitchy-jsonp) [inline tpm/tpm.go]: Ditto: Should be okay to `return nil` here, as the error case is already handled above.

File: tpm/tpm.go [hunk 5]
```diff
@@ -940,6 +940,39 @@ func NVReadValue(rw io.ReadWriter, index, offset, len uint32, ownAuth []byte) ([
...
 
+// NVWriteValue for writing to the NVRAM. Needs a index for a defined space in NVRAM.
+// See TPM-Main-Part-3-Commands_v1.2_rev116_01032011, P216
+func NVWriteValue(rw io.ReadWriter, index, offset uint32, data []byte, ownAuth []byte) error {
+	if ownAuth == nil {
+		_, _, _, err := nvWriteValue(rw, index, offset, uint32(len(data)), data, nil)
+		if err != nil {
+			return fmt.Errorf("failed to read from NVRAM: %v", err)
+		}
+		return err
```
Comment (twitchy-jsonp) [inline tpm/tpm.go]: Should be okay to `return nil` here, as the error case is already handled above.

File: tpm/tpm_test.go [hunk 6]
```diff
@@ -538,6 +540,11 @@ func TestOwnerReadSRK(t *testing.T) {
 	}
 }
 
+func TestNVWrite(t *testing.T) {
```
Comment (twitchy-jsonp) [inline tpm/tpm_test.go]: Do you need this test stub now theres `TestNVWriteValue` ?

File: tpm/tpm.go (hunk 5 above)
Comment (twitchy-jsonp) [reply tpm/tpm.go]: One idea:

```go
if _, _, _, err := nvWriteValue(rw, index, offset, uint32(len(data)), data, nil); err != nil {
	return fmt.Errorf("failed to read from NVRAM: %v", err)
}
return nil
```

File: tpm/tpm_test.go [hunk 7]
```diff
@@ -604,3 +606,33 @@ func TestForceClear(t *testing.T) {
...
+	}
+
+	indexValue, err := NVReadValue(rwc, rndIndex, 0, rndIndexInfo.Size, nil)
+	if err != nil {
+		t.Fatalf("Couldn't read index value @ index: %v with error: %v", rndIndex, err)
+	}
+
+	err = NVWriteValue(rwc, rndIndex, 0, indexValue, nil)
+	if err != nil {
+		t.Fatalf("Couldn't write to index: %v with error: %v", rndIndex, err)
```
Comment (twitchy-jsonp) [inline tpm/tpm_test.go]: Use `t.Errorf` if the test can continue or you are doing the final check in the test function.

File: tpm/tpm.go [hunk 8]
```diff
@@ -940,6 +940,38 @@ func NVReadValue(rw io.ReadWriter, index, offset, len uint32, ownAuth []byte) ([
...
+	defer osaprOwn.Close(rw)
+	defer zeroBytes(sharedSecretOwn[:])
+	authIn := []interface{}{ordNVWriteValue, index, offset, len(data), data}
+	ca, err := newCommandAuth(osaprOwn.AuthHandle, osaprOwn.NonceEven, sharedSecretOwn[:], authIn)
+	if err != nil {
+		return fmt.Errorf("failed to construct owner auth fields: %v", err)
+	}
+	data, ra, ret, err := nvWriteValue(rw, index, offset, uint32(len(data)), data, ca)
+	if err != nil {
+		return fmt.Errorf("failed to read from NVRAM: %v", err)
```
Comment (twitchy-jsonp) [inline tpm/tpm.go]: should this read: failed to _write_ from NVRAM?

File: tpm/tpm_test.go [hunk 9]
```diff
@@ -604,3 +606,33 @@ func TestForceClear(t *testing.T) {
 		t.Fatal("Couldn't clear the TPM without owner auth in physical presence mode:", err)
 	}
 }
+
+func TestNVWriteValue(t *testing.T) {
+	mathrand.Seed(time.Now().Unix())
```
Comment (twitchy-jsonp) [inline tpm/tpm_test.go]: Nit: Its usually good to keep tests as deterministic as possible. As such, consider using a fixed seed: `mathrand.Seed(42)`

**Comment by twitchy-jsonp** [comment]
@ChriMarMe you might need to merge in the latest master from go-tpm before it will let us merge this.

@tmroeder Github has you are marked as a reviewer, do you have outstanding comments / review?

File: tpm/tpm.go [hunk 10]
```diff
@@ -981,6 +981,38 @@ func NVReadValue(rw io.ReadWriter, index, offset, len uint32, ownAuth []byte) ([
 	return data, nil
 }
 
+// NVWriteValue for writing to the NVRAM. Needs a index for a defined space in NVRAM.
+// See TPM-Main-Part-3-Commands_v1.2_rev116_01032011, P216
+func NVWriteValue(rw io.ReadWriter, index, offset uint32, data []byte, ownAuth []byte) error {
+	if ownAuth == nil {
+		if _, _, _, err := nvWriteValue(rw, index, offset, uint32(len(data)), data, nil); err != nil {
+			return fmt.Errorf("failed to read from NVRAM: %v", err)
```
Comment (twitchy-jsonp) [inline tpm/tpm.go]: This should read `failed to write from NVRAM`

File: tpm/tpm.go [hunk 11]
```diff
@@ -981,6 +981,38 @@ func NVReadValue(rw io.ReadWriter, index, offset, len uint32, ownAuth []byte) ([
...
+	defer osaprOwn.Close(rw)
+	defer zeroBytes(sharedSecretOwn[:])
+	authIn := []interface{}{ordNVWriteValue, index, offset, len(data), data}
+	ca, err := newCommandAuth(osaprOwn.AuthHandle, osaprOwn.NonceEven, sharedSecretOwn[:], authIn)
+	if err != nil {
+		return fmt.Errorf("failed to construct owner auth fields: %v", err)
+	}
+	data, ra, ret, err := nvWriteValue(rw, index, offset, uint32(len(data)), data, ca)
+	if err != nil {
+		return fmt.Errorf("failed to write from NVRAM: %v", err)
```
Comment (twitchy-jsonp) [inline tpm/tpm.go]: This should read: `failed to write to NVRAM`

**Comment by ChriMarMe** [comment]
Took me some time. I hope this will get it.

**Pull request merged**